interpreter/cling/test/DynamicLibraryManager/cached_realpath.C

```diff
@@ -71,19 +71,20 @@
 #include "../lib/Interpreter/DynamicLibraryManagerSymbol.cpp"
 
 .rawInput 1
-void test_realpath(std::string path) {
+void test_realpath(const std::string &path) {
+  int err_s = 0, err_c = 0;
+
   // system realpath
-  errno = 0;
-  char system_resolved_path[4096];
-  system_resolved_path[0] = '\0';
-  realpath(path.c_str(), system_resolved_path);
-  int err_s = errno;
-  if (err_s !=0 ) system_resolved_path[0] = '\0';
+  char system_resolved_path[4096]{};
+  if (!realpath(path.c_str(), system_resolved_path)) {
+    system_resolved_path[0] = '\0';
+    err_s = errno;
+  }
```

Comment:
Hm, I thought setting `errno = 0` before a stdlib call was the standard way of doing it... So you are saying that `realpath` can be successful and still set a non-zero `errno`?

Reply:
Indeed, at least for the glibc implementation (see https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/canonicalize.c;h=7e421ef11559049e8257f78cbc3e85474fd62553;hb=HEAD#l188).

Specifically, `realpath()` issues a number of `readlink()` syscalls that might return `EINVAL`, apparently due to "The  named file (i.e., the final filename component of pathname) is not a symbolic link." (those are visible using `strace`).
However, the [normal return path](https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/canonicalize.c;h=7e421ef11559049e8257f78cbc3e85474fd62553;hb=HEAD#l398) does not reset `errno` (that may leak from former `readlink()` calls).

In principle this is still consistent with the manual page for `realpath(3)`, i.e. it does not make any guarantee on the value of `errno` for a successful call:
```
If  there  is  no  error,  realpath()  returns  a  pointer  to  the resolved_path.
Otherwise, it returns NULL, the contents of the array resolved_path are undefined, and errno is set to indicate the error.
```